File: mcp/src/runsheet.ts
/**
 * The runsheet an agent writes, checked against the app's own model (src/features/runsheet/model.ts).
 * Every enum below is keyed by the model's type, so adding a mode to the app without adding it here
 * fails the type-check; `normalise` returns a value typed as the model's Runsheet.
 */
import { z } from 'zod';
import type { Block, BlockMode, ExerciseRef, ExerciseStep, ForMode, Item, ItemRole, RestStep, Runsheet, ScoreType, Step } from '@/features/runsheet/model';
import { runsheetMinutes, uid } from '@/features/runsheet/model';

const FOR_MODE: Record<ForMode, string> = {
  reps: 'forValue reps',
  seconds: 'forValue seconds of work (the timer counts down)',
  minutes: 'forValue minutes (steady cardio)',
  meters: 'forValue metres',
  calories: 'forValue calories (rower, bike)',
  max: 'as many reps / as long as possible; forValue ignored',
  amrap: 'at least forValue reps, then as many as possible ("5+")',
  segment: 'a stretch of a follow-along video, startSeconds to endSeconds',
};
const BLOCK_MODE: Record<BlockMode, string> = {
  rounds: 'repeat the steps `repeat` times',
  fortime: '`repeat` rounds as fast as possible, clock counts up, optional timeCapSec',
  amrap: 'as many rounds as possible in timeCapSec (required)',
  emom: 'start the steps every everySec seconds (default 60), `repeat` times',
  ladder: 'one pass per rung of `ladder` (e.g. [21,15,9]); rep steps take the rung as their reps',
};
const SCORE: Record<ScoreType, true> = { time: true, rounds: true, reps: true, load: true, distance: true, none: true };
const ROLE: Record<ItemRole, true> = { warmup: true, main: true, cooldown: true };

const keys = <K extends string>(r: Record<K, unknown>) => Object.keys(r) as [K, ...K[]];
const forMode = z.enum(keys(FOR_MODE));
const blockMode = z.enum(keys(BLOCK_MODE));
const score = z.enum(keys(SCORE));
const role = z.enum(keys(ROLE)).describe('warmup | main (default) | cooldown; the app draws dividers between parts');
const num = z.number().finite();
const pos = num.nonnegative();

const exerciseRef = z.object({
  key: z.string().min(1).max(80),
  name: z.string().min(1).max(80),
  unit: z.string().max(20).describe('Unit of `target`: "kg", "kg per arm", "kph", or "" for bodyweight'),
  step: num.positive().describe('Stepper increment for target, e.g. 2.5'),
  cue: z.string().max(200).optional(),
});

const progression = z.object({ onSuccessKg: pos.optional(), deloadPct: pos.max(100).optional(), failAfter: z.number().int().positive().optional(), amrapBumpAt: z.number().int().positive().optional(), tmBumpKg: pos.optional() });

export const exerciseStepInput = z.object({
  kind: z.literal('exercise'),
  id: z.string().max(60).optional(),
  exercise: z.union([z.string().min(1), exerciseRef]).describe('An exercise key from search_exercises (preferred), or a full {key,name,unit,step} for something not in the library'),
  forMode,
  forValue: pos.describe('Reps, seconds, minutes, metres or calories, per forMode'),
  forMax: pos.optional().describe('Upper bound of a range: 8–12 reps is forValue 8, forMax 12'),
  target: pos.optional().describe("Load or speed in the exercise's unit (kg, kg per arm, kph). Omit for bodyweight"),
  perSide: z.boolean().optional().describe('Reps/time are per side; the timer doubles it'),
  loadFactor: pos.optional().describe('Load as a multiple of bodyweight, instead of target'),
  targetPct: pos.max(200).optional().describe('Load as % of the training max, instead of target'),
  incline: pos.max(40).optional().describe('Treadmill incline %'),
  rx: z.object({ men: pos.optional(), women: pos.optional() }).optional(),
  startSeconds: pos.optional(),
  endSeconds: pos.optional(),
  ladderFactor: pos.optional(),
  ladderFixed: z.boolean().optional(),
  role: role.optional(),
  note: z.string().max(300).optional(),
});
export const restStepInput = z.object({ kind: z.literal('rest'), id: z.string().max(60).optional(), seconds: num.int().positive().max(3600), role: role.optional(), note: z.string().max(300).optional() });
const stepInput = z.discriminatedUnion('kind', [exerciseStepInput, restStepInput]);
export const blockInput = z.object({
  kind: z.literal('block'),
  id: z.string().max(60).optional(),
  name: z.string().max(80).optional().describe('Defaults to the exercise names joined with " + "'),
  repeat: num.int().positive().max(100).describe('Rounds; ignored for amrap'),
  mode: blockMode.optional().describe(Object.entries(BLOCK_MODE).map(([k, v]) => `${k}: ${v}`).join('; ')),
  timeCapSec: num.int().positive().optional(),
  everySec: num.int().positive().optional(),
  ladder: z.array(num.int().positive()).max(50).optional(),
  restBetweenSec: num.int().nonnegative().optional().describe('Rest between repeats of the block, not after the last'),
  score: score.optional(),
  progression: progression.optional(),
  role: role.optional(),
  note: z.string().max(300).optional(),
  steps: z.array(stepInput).min(1).max(40),
});
const itemInput = z.discriminatedUnion('kind', [exerciseStepInput, restStepInput, blockInput]);

export const runsheetInput = z.object({
  title: z.string().trim().min(1).max(120),
  description: z.string().max(2000).optional(),
  creator: z.string().max(80).optional().describe("Shown as the author; defaults to the user's name"),
  tags: z.array(z.string().max(30)).max(12).optional(),
  level: z.enum(['Easy', 'Medium', 'Hard']).optional(),
  timeCapSec: num.int().positive().optional().describe('One cap across the whole workout'),
  score: score.optional(),
  progression: progression.optional(),
  items: z.array(itemInput).min(1).max(60),
});
export type RunsheetInput = z.infer<typeof runsheetInput>;

export const FOR_MODE_HELP = Object.entries(FOR_MODE).map(([k, v]) => `${k} = ${v}`).join('; ');

export class RunsheetError extends Error {}

/**
 * Input → the model's Runsheet: exercise keys become full ExerciseRefs from `library`, missing ids
 * are filled, and the rules the timer relies on are checked. Throws RunsheetError with every
 * problem listed, so an agent can fix them all in one go.
 */
export const normalise = (input: RunsheetInput, library: (key: string) => ExerciseRef | undefined): Runsheet => {
  const problems: string[] = [];
  const seen = new Set<string>();
  const id = (given: string | undefined, prefix: string) => {
    let v = given && !seen.has(given) ? given : uid(prefix);
    while (seen.has(v)) v = uid(prefix);
    seen.add(v);
    return v;
  };
  const step = (s: z.infer<typeof stepInput>, where: string): Step => {
    if (s.kind === 'rest') return { ...s, id: id(s.id, 'r') } satisfies RestStep;
    let exercise: ExerciseRef | undefined;
    if (typeof s.exercise === 'string') {
      exercise = library(s.exercise);
      if (!exercise) problems.push(`${where}: unknown exercise key "${s.exercise}" — find one with search_exercises, or pass {key,name,unit,step}`);
    } else exercise = library(s.exercise.key) ?? s.exercise;
    if (s.forMode === 'segment' && s.startSeconds === undefined) problems.push(`${where}: forMode "segment" needs startSeconds`);
    if (s.forMax !== undefined && s.forMax < s.forValue) problems.push(`${where}: forMax (${s.forMax}) is below forValue (${s.forValue})`);
    if (s.target !== undefined && exercise && exercise.unit === '') problems.push(`${where}: "${exercise.name}" is bodyweight, so it takes no target`);
    const out: ExerciseStep = { ...s, id: id(s.id, 'e'), exercise: exercise ?? { key: String(s.exercise), name: String(s.exercise), unit: '', step: 1 } };
    return out;
  };
  const items: Item[] = input.items.map((it, i) => {
    const where = `items[${i}]`;
    if (it.kind !== 'block') return step(it, where);
    const mode = it.mode ?? 'rounds';
    if (mode === 'amrap' && !it.timeCapSec) problems.push(`${where}: an amrap block needs timeCapSec`);
    if (mode === 'ladder' && !it.ladder?.length) problems.push(`${where}: a ladder block needs ladder, e.g. [21,15,9]`);
    const steps = it.steps.map((s, j) => step(s, `${where}.steps[${j}]`));
    const names = [...new Set(steps.filter((s): s is ExerciseStep => s.kind === 'exercise').map(s => s.exercise.name))];
    const block: Block = { ...it, id: id(it.id, 'b'), name: it.name?.trim() || names.join(' + ') || 'Block', steps };
    return block;
  });
  if (problems.length) throw new RunsheetError(problems.join('\n'));
  const r: Runsheet = { ...input, items };
  if (runsheetMinutes(r) > 300) throw new RunsheetError(`That runs for about ${runsheetMinutes(r)} minutes; keep a workout under 5 hours`);
  return r;
};

File: ios/TigerWorkouts/Model/Runsheet.swift
import Foundation

/// The runsheet model, ported from `src/features/runsheet/model.ts`. A workout is an ordered list
/// of items; an item is a step (exercise or rest) or a block of steps that repeats N times.
/// Everything here is a value type, so the engine can hand new copies back up the way the TS does.

enum ForMode: String, Codable, Sendable {
    case seconds, reps, minutes, meters, calories, max, amrap, segment
}

enum ScoreType: String, Codable, Sendable {
    case time, rounds, reps, load, distance, none
}

enum ItemRole: String, Codable, Sendable {
    case warmup, main, cooldown

    var label: String {
        switch self {
        case .warmup: "Warm-up"
        case .main: "Workout"
        case .cooldown: "Cool-down"
        }
    }
}

enum BlockMode: String, Codable, Sendable {
    case rounds, fortime, amrap, emom, ladder
}

struct Rx: Codable, Hashable, Sendable {
    var men: Double?
    var women: Double?
}

struct Progression: Codable, Hashable, Sendable {
    var onSuccessKg: Double?
    var deloadPct: Double?
    var failAfter: Int?
    var amrapBumpAt: Int?
    var tmBumpKg: Double?
}

struct Source: Codable, Hashable, Sendable {
    var title: String = ""
    var url: String?
    var author: String?
    var kind: String = "user"
    var license: String?
    var importedAt: String?
}

struct Video: Codable, Hashable, Sendable {
    var provider: String?
    var id: String?
    var url: String?
}

struct ProgramRef: Codable, Hashable, Sendable {
    var name: String = ""
    var day: String = ""
    var order: Int?
}

/// One exercise as the library knows it. The export writes `exercise` as a key string; runsheets
/// written by the web app carry the whole object inline, so both decode.
struct ExerciseRef: Codable, Hashable, Sendable, Identifiable {
    var key: String
    var name: String
    var unit: String
    var step: Double
    var clip: String?
    var poster: String?
    var icon: String?
    var cue: String?

    var id: String { key }

    /// A stand-in for a key no library knows: "bb_thruster" reads as "thruster".
    static func placeholder(key: String) -> ExerciseRef {
        let name = key.replacingOccurrences(of: "^[a-z]+_", with: "", options: .regularExpression)
            .replacingOccurrences(of: "_", with: " ")
        return ExerciseRef(key: key, name: name.isEmpty ? key : name, unit: "", step: 1)
    }

    init(key: String, name: String, unit: String, step: Double, clip: String? = nil, poster: String? = nil, icon: String? = nil, cue: String? = nil) {
        self.key = key
        self.name = name
        self.unit = unit
        self.step = step
        self.clip = clip
        self.poster = poster
        self.icon = icon
        self.cue = cue
    }

    init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        key = try c.decodeIfPresent(String.self, forKey: .key) ?? ""
        name = try c.decodeIfPresent(String.self, forKey: .name) ?? ExerciseRef.placeholder(key: key).name
        unit = try c.decodeIfPresent(String.self, forKey: .unit) ?? ""
        step = try c.decodeIfPresent(Double.self, forKey: .step) ?? 1
        clip = try c.decodeIfPresent(String.self, forKey: .clip)
        poster = try c.decodeIfPresent(String.self, forKey: .poster)
        icon = try c.decodeIfPresent(String.self, forKey: .icon)
        cue = try c.decodeIfPresent(String.self, forKey: .cue)
    }
}

struct ExerciseStep: Codable, Hashable, Sendable, Identifiable {
    var id: String
    var exercise: ExerciseRef
    var target: Double?
    var forMode: ForMode = .seconds
    var forValue: Double = 30
    var forMax: Double?
    var perSide: Bool?
    var loadFactor: Double?
    var targetPct: Double?
    var incline: Double?
    var rx: Rx?
    var startSeconds: Double?
    var endSeconds: Double?
    var ladderFactor: Double?
    var ladderFixed: Bool?
    var role: ItemRole?
    var note: String?

    private enum CodingKeys: String, CodingKey {
        case id, exercise, target, forMode, forValue, forMax, perSide, loadFactor, targetPct
        case incline, rx, startSeconds, endSeconds, ladderFactor, ladderFixed, role, note
    }

    init(id: String, exercise: ExerciseRef, target: Double? = nil, forMode: ForMode = .seconds, forValue: Double = 30, incline: Double? = nil, perSide: Bool? = nil) {
        self.id = id
        self.exercise = exercise
        self.target = target
        self.forMode = forMode
        self.forValue = forValue
        self.incline = incline
        self.perSide = perSide
    }

    init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(String.self, forKey: .id) ?? UUID().uuidString
        // A key string resolves against the bundled library; an inline object is used as written.
        if let key = try? c.decode(String.self, forKey: .exercise) {
            exercise = decoder.exerciseLibrary?[key] ?? ExerciseRef.placeholder(key: key)
        } else {
            exercise = try c.decode(ExerciseRef.self, forKey: .exercise)
        }
        target = try c.decodeIfPresent(Double.self, forKey: .target)
        forMode = try c.decodeIfPresent(ForMode.self, forKey: .forMode) ?? .seconds
        forValue = try c.decodeIfPresent(Double.self, forKey: .forValue) ?? 30
        forMax = try c.decodeIfPresent(Double.self, forKey: .forMax)
        perSide = try c.decodeIfPresent(Bool.self, forKey: .perSide)
        loadFactor = try c.decodeIfPresent(Double.self, forKey: .loadFactor)
        targetPct = try c.decodeIfPresent(Double.self, forKey: .targetPct)
        incline = try c.decodeIfPresent(Double.self, forKey: .incline)
        rx = try c.decodeIfPresent(Rx.self, forKey: .rx)
        startSeconds = try c.decodeIfPresent(Double.self, forKey: .startSeconds)
        endSeconds = try c.decodeIfPresent(Double.self, forKey: .endSeconds)
        ladderFactor = try c.decodeIfPresent(Double.self, forKey: .ladderFactor)
        ladderFixed = try c.decodeIfPresent(Bool.self, forKey: .ladderFixed)
        role = try c.decodeIfPresent(ItemRole.self, forKey: .role)
        note = try c.decodeIfPresent(String.self, forKey: .note)
    }

    func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encode(exercise, forKey: .exercise)
        try c.encodeIfPresent(target, forKey: .target)
        try c.encode(forMode, forKey: .forMode)
        try c.encode(forValue, forKey: .forValue)
        try c.encodeIfPresent(forMax, forKey: .forMax)
        try c.encodeIfPresent(perSide, forKey: .perSide)
        try c.encodeIfPresent(loadFactor, forKey: .loadFactor)
        try c.encodeIfPresent(targetPct, forKey: .targetPct)
        try c.encodeIfPresent(incline, forKey: .incline)
        try c.encodeIfPresent(rx, forKey: .rx)
        try c.encodeIfPresent(startSeconds, forKey: .startSeconds)
        try c.encodeIfPresent(endSeconds, forKey: .endSeconds)
        try c.encodeIfPresent(ladderFactor, forKey: .ladderFactor)
        try c.encodeIfPresent(ladderFixed, forKey: .ladderFixed)
        try c.encodeIfPresent(role, forKey: .role)
        try c.encodeIfPresent(note, forKey: .note)
    }
}

struct RestStep: Codable, Hashable, Sendable, Identifiable {
    var id: String
    var seconds: Double
    var role: ItemRole?
    var note: String?

    init(id: String, seconds: Double, role: ItemRole? = nil, note: String? = nil) {
        self.id = id
        self.seconds = seconds
        self.role = role
        self.note = note
    }

    init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(String.self, forKey: .id) ?? UUID().uuidString
        seconds = try c.decodeIfPresent(Double.self, forKey: .seconds) ?? 30
        role = try c.decodeIfPresent(ItemRole.self, forKey: .role)
        note = try c.decodeIfPresent(String.self, forKey: .note)
    }
}

enum Step: Codable, Hashable, Sendable, Identifiable {
    case exercise(ExerciseStep)
    case rest(RestStep)

    var id: String {
        switch self {
        case .exercise(let s): s.id
        case .rest(let s): s.id
        }
    }

    var asExercise: ExerciseStep? {
        if case .exercise(let s) = self { return s }
        return nil
    }

    var role: ItemRole? {
        switch self {
        case .exercise(let s): s.role
        case .rest(let s): s.role
        }
    }

    private enum Discriminator: String, CodingKey { case kind }

    init(from decoder: any Decoder) throws {
        let k = try decoder.container(keyedBy: Discriminator.self)
        if try k.decodeIfPresent(String.self, forKey: .kind) == "rest" {
            self = .rest(try RestStep(from: decoder))
        } else {
            self = .exercise(try ExerciseStep(from: decoder))
        }
    }

    func encode(to encoder: any Encoder) throws {
        var k = encoder.container(keyedBy: Discriminator.self)
        switch self {
        case .exercise(let s):
            try k.encode("exercise", forKey: .kind)
            try s.encode(to: encoder)
        case .rest(let s):
            try k.encode("rest", forKey: .kind)
            try s.encode(to: encoder)
        }
    }
}

struct Block: Codable, Hashable, Sendable, Identifiable {
    var id: String
    var name: String
    /// Rounds for rounds / fortime / emom. Ignored for amrap. `repeat` in the JSON.
    var repeatCount: Int = 1
    var mode: BlockMode?
    var timeCapSec: Double?
    var everySec: Double?
    var ladder: [Double]?
    var restBetweenSec: Double?
    var score: ScoreType?
    var progression: Progression?
    var role: ItemRole?
    var note: String?
    var steps: [Step] = []

    var runMode: BlockMode { mode ?? .rounds }

    private enum CodingKeys: String, CodingKey {
        case id, name, mode, timeCapSec, everySec, ladder, restBetweenSec, score, progression, role, note, steps
        case repeatCount = "repeat"
    }

    init(id: String, name: String, repeatCount: Int = 1, mode: BlockMode? = nil, steps: [Step] = [], restBetweenSec: Double? = nil, timeCapSec: Double? = nil, everySec: Double? = nil, ladder: [Double]? = nil) {
        self.id = id
        self.name = name
        self.repeatCount = repeatCount
        self.mode = mode
        self.steps = steps
        self.restBetweenSec = restBetweenSec
        self.timeCapSec = timeCapSec
        self.everySec = everySec
        self.ladder = ladder
    }

    init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(String.self, forKey: .id) ?? UUID().uuidString
        name = try c.decodeIfPresent(String.self, forKey: .name) ?? "Block"
        repeatCount = try c.decodeIfPresent(Int.self, forKey: .repeatCount) ?? 1
        mode = try c.decodeIfPresent(BlockMode.self, forKey: .mode)
        timeCapSec = try c.decodeIfPresent(Double.self, forKey: .timeCapSec)
        everySec = try c.decodeIfPresent(Double.self, forKey: .everySec)
        ladder = try c.decodeIfPresent([Double].self, forKey: .ladder)
        restBetweenSec = try c.decodeIfPresent(Double.self, forKey: .restBetweenSec)
        score = try c.decodeIfPresent(ScoreType.self, forKey: .score)
        progression = try c.decodeIfPresent(Progression.self, forKey: .progression)
        role = try c.decodeIfPresent(ItemRole.self, forKey: .role)
        note = try c.decodeIfPresent(String.self, forKey: .note)
        steps = try c.decodeIfPresent([Step].self, forKey: .steps) ?? []
    }
}

struct RefItem: Codable, Hashable, Sendable, Identifiable {
    var id: String
    var runsheetId: String
    var role: ItemRole?
}

enum Item: Codable, Hashable, Sendable, Identifiable {
    case step(Step)
    case block(Block)
    case ref(RefItem)

    var id: String {
        switch self {
        case .step(let s): s.id
        case .block(let b): b.id
        case .ref(let r): r.id
        }
    }

    var asBlock: Block? {
        if case .block(let b) = self { return b }
        return nil
    }

    var asStep: Step? {
        if case .step(let s) = self { return s }
        return nil
    }

    var role: ItemRole? {
        switch self {
        case .step(let s): s.role
        case .block(let b): b.role
        case .ref(let r): r.role
        }
    }

    private enum Discriminator: String, CodingKey { case kind }

    init(from decoder: any Decoder) throws {
        let k = try decoder.container(keyedBy: Discriminator.self)
        switch try k.decodeIfPresent(String.self, forKey: .kind) {
        case "block": self = .block(try Block(from: decoder))
        case "ref": self = .ref(try RefItem(from: decoder))
        default: self = .step(try Step(from: decoder))
        }
    }

    func encode(to encoder: any Encoder) throws {
        switch self {
        case .step(let s):
            try s.encode(to: encoder)
        case .block(let b):
            var k = encoder.container(keyedBy: Discriminator.self)
            try k.encode("block", forKey: .kind)
            try b.encode(to: encoder)
        case .ref(let r):
            var k = encoder.container(keyedBy: Discriminator.self)
            try k.encode("ref", forKey: .kind)
            try r.encode(to: encoder)
        }
    }
}

struct Runsheet: Codable, Hashable, Sendable, Identifiable {
    var id: String?
    var title: String
    var creator: String?
    var description: String?
    var source: Source?
    var tags: [String]?
    var level: String?
    var program: ProgramRef?
    var timeCapSec: Double?
    var score: ScoreType?
    var progression: Progression?
    var video: Video?
    /// Your own workouts only: true shows it in everyone's Discover. New ones start private.
    /// `public` in the JSON, as the web app writes it.
    var isPublic: Bool?
    /// Set on your own version of someone else's workout: the id it was made from.
    var derivedFrom: String?
    var items: [Item] = []

    /// Stable identity for lists: the id if it has one, else the title.
    var key: String { id ?? title }

    private enum CodingKeys: String, CodingKey {
        case id, title, creator, description, source, tags, level, program, timeCapSec, score, progression, video, derivedFrom, items
        case isPublic = "public"
    }

    init(id: String? = nil, title: String, creator: String? = nil, items: [Item] = [], timeCapSec: Double? = nil, score: ScoreType? = nil) {
        self.id = id
        self.title = title
        self.creator = creator
        self.items = items
        self.timeCapSec = timeCapSec
        self.score = score
    }

    init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decodeIfPresent(String.self, forKey: .id)
        title = try c.decodeIfPresent(String.self, forKey: .title) ?? "Workout"
        creator = try c.decodeIfPresent(String.self, forKey: .creator)
        description = try c.decodeIfPresent(String.self, forKey: .description)
        source = try c.decodeIfPresent(Source.self, forKey: .source)
        tags = try c.decodeIfPresent([String].self, forKey: .tags)
        level = try c.decodeIfPresent(String.self, forKey: .level)
        program = try c.decodeIfPresent(ProgramRef.self, forKey: .program)
        timeCapSec = try c.decodeIfPresent(Double.self, forKey: .timeCapSec)
        score = try c.decodeIfPresent(ScoreType.self, forKey: .score)
        progression = try c.decodeIfPresent(Progression.self, forKey: .progression)
        video = try c.decodeIfPresent(Video.self, forKey: .video)
        isPublic = try c.decodeIfPresent(Bool.self, forKey: .isPublic)
        derivedFrom = try c.decodeIfPresent(String.self, forKey: .derivedFrom)
        items = try c.decodeIfPresent([Item].self, forKey: .items) ?? []
    }
}

// MARK: - Library injection

extension CodingUserInfoKey {
    static let exerciseLibrary = CodingUserInfoKey(rawValue: "tiger.exerciseLibrary")!
}

extension Decoder {
    var exerciseLibrary: [String: ExerciseRef]? {
        userInfo[.exerciseLibrary] as? [String: ExerciseRef]
    }
}
